### ISSUE TITLE
[Failing Test] [sig-scheduling] [Feature:GPUDevicePlugin] run Nvidia GPU Device Plugin tests (ci-kubernetes-e2e-gce-device-plugin-gpu)

### ISSUE THREAD

**Issue by thejoycekung** [kubernetes/kubernetes#96519]
**Which jobs are failing**:
`ci-kubernetes-e2e-gce-device-plugin-gpu`

**Which test(s) are failing**:
`[sig-scheduling] [Feature:GPUDevicePlugin] run Nvidia GPU Device Plugin tests`

**Since when has it been failing**:
https://github.com/kubernetes/kubernetes/compare/0469db9fe...dd4560370

**Testgrid link**:
https://testgrid.k8s.io/sig-release-master-blocking#gce-device-plugin-gpu-master

**Reason for failure**:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/scheduling/nvidia-gpus.go:224
Nov 11 21:43:04.675: wait for pod "nvidia-gpu-13237348-f464-496a-9e10-7d2bd90a5aad" to succeed
Expected success, but got an error:
    <*errors.errorString | 0xc003262610>: {
        s: "pod \"nvidia-gpu-13237348-f464-496a-9e10-7d2bd90a5aad\" failed with reason: \"UnexpectedAdmissionError\", message: \"Pod Allocate failed due to requested number of devices unavailable for nvidia.com/gpu. Requested: 1, Available: 0, which is unexpected\"",
    }
    pod "nvidia-gpu-13237348-f464-496a-9e10-7d2bd90a5aad" failed with reason: "UnexpectedAdmissionError", message: "Pod Allocate failed due to requested number of devices unavailable for nvidia.com/gpu. Requested: 1, Available: 0, which is unexpected"
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/pods.go:212
```

**Anything else we need to know**:
Can't find this in Triage

Example spyglass links: 
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-device-plugin-gpu/1326638573817958400
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-device-plugin-gpu/1326729172260753408
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-device-plugin-gpu/1326698973720743936
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-device-plugin-gpu/1326668771879817216

/sig scheduling
/cc @kubernetes/ci-signal @kubernetes/sig-scheduling-bugs 
/priority important-soon

**Comment by Huang-Wei** [comment]
@ahg-g I don't quite think this test should be owned by SIG-Scheduling. Most of the failures happen in the Nvidia GPU setup phase.

BTW: although irrelevant, last time we see a failure of this test is #93187.

**Comment by Huang-Wei** [comment]
Searching keyword "UnexpectedAdmissionError" gets the following occurrences:

```
pkg/kubelet/lifecycle/predicate.go
82:			Reason:  "UnexpectedAdmissionError",
104:			Reason:  "UnexpectedAdmissionError",
116:				Reason:  "UnexpectedAdmissionError",

pkg/kubelet/cm/container_manager_linux.go
724:				Reason:  "UnexpectedAdmissionError",
734:					Reason:  "UnexpectedAdmissionError",

pkg/kubelet/cm/topologymanager/topology_manager.go
242:					Reason:  "UnexpectedAdmissionError",
268:				Reason:  "UnexpectedAdmissionError",
```

**Comment by Huang-Wei** [comment]
/sig node

cc @kubernetes/sig-node-bugs in case you're aware of any recent changes that may have caused this failure.

**Comment by ruiwen-zhao** [comment]
From the [test log](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-device-plugin-gpu/1326940321854001152) I can see a bunch of:

```
I1112 17:37:35.700] Nov 12 17:37:35.700: INFO: gpuResourceName nvidia.com/gpu
I1112 17:37:35.701] Nov 12 17:37:35.700: INFO: Nvidia GPUs not available on Node: "bootstrap-e2e-minion-group-2hjf"
```

So it looks like the device plugin on bootstrap-e2e-minion-group-2hjf did not discover any GPU.

**Comment by ruiwen-zhao** [comment]
The nil pointer is from here: https://github.com/kubernetes/kubernetes/blob/6d5cb36d36f34cb4f5735b6adcd5ea8ebb4440ba/pkg/kubelet/cm/devicemanager/manager.go#L908

, which is a new change introduced by https://github.com/kubernetes/kubernetes/commit/a8b8995ef241e93e9486d475126450f33f24ef4e

**Comment by ruiwen-zhao** [comment]
The current device plugin (the one defined [here](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/device-plugins/nvidia-gpu/daemonset.yaml#L39)) is using an old version of device plugin api (see [here](https://github.com/GoogleCloudPlatform/container-engine-accelerators/blob/e3cce9a52ea0034e94512a470f2f26b672204816/vendor/k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1/api.proto)), which does not have `Topology` field under `Device`. So calling `m.allDevices[resource][dev].Topology.Nodes` will end up with NPE.

**Comment by Huang-Wei** [comment]
Thanks @ruiwen-zhao , the analysis makes great sense to me. Could you please work on a fix to address this? (Probably updating https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/device-plugins/nvidia-gpu/daemonset.yaml)

**Comment by ruiwen-zhao** [comment]
Hi @Huang-Wei , I am still looking at the code, because the code does have the logic to handle the case where device plugin does not return affinity: https://github.com/kubernetes/kubernetes/blob/6d5cb36d36f34cb4f5735b6adcd5ea8ebb4440ba/pkg/kubelet/cm/devicemanager/manager.go#L745

Also, even if the Topology is not defined, I am not sure `m.allDevices[resource][dev].Topology.Nodes` will throw NPE (https://play.golang.org/p/Fni4TMFxe1o) (please correct me if I am wrong here). I will need to look more closely at the code to find out where the NPE is from.

Updating the https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/device-plugins/nvidia-gpu/daemonset.yaml might take long time because we will need some testing to make sure it doesn't break anything. Currently I am thinking of identifying the source of NPE, and make sure `manager.go` can handle that case.

**Comment by AlexeyPerevalov** [comment]
I submitted a fix. My fault, it was a real case when pointer to Topology could be a nil.
But    still fails:
I don't observe SIGSEGV now in kubelet.logs, but 2 problems there:
1. F1113 09:26:19.369703    1336 server.go:269] failed to run Kubelet: cannot create certificate signing request: certificatesigningrequests.certificates.k8s.io is forbidden: User "kubelet" cannot create resource "certificatesigningrequests" in API group "certificates.k8s.io" at the cluster scope
2. pod of nvidia-gpu stops and restarts after a while.

**Comment by ruiwen-zhao** [comment]
@AlexeyPerevalov can you show full error msg of "NVIDIA GPU metrics will not be available: %s"?

**Comment by AlexeyPerevalov** [comment]
> NVIDIA GPU metrics will not be available

I1113 15:43:26.937884    1338 nvidia.go:95] Found device with vendorID "0x10de"
W1113 15:43:26.938084    1338 nvidia.go:61] **NVIDIA GPU metrics will not be available**: Could not initialize NVML: could not load NVML library
I1113 15:43:26.950281    1338 manager.go:213] Machine: {Timestamp:2020-11-13 15:43:26.949899613 +0000 UTC m=+5.925633071 NumCores:2 NumPhysicalCores:1 NumSockets:1 CpuFrequency:2200000 MemoryCapacity:7823970304 MemoryByType:map[] NVMInfo:{MemoryModeCapacity:0 AppDirectModeCapacity:0 AvgPowerBudget:0} HugePages:[{PageSize:2048 NumPages:0}] MachineID:9542199b98243a64f41f91f95568c08d SystemUUID:9542199b-9824-3a64-f41f-91f95568c08d BootID:d43ff90c-f891-4c0b-a34d-eb5812b906e0 Filesystems:[{Device:/tmp DeviceMajor:0 DeviceMinor:39 Capacity:3911983104 Type:vfs Inodes:955074 HasInodes:true} {Device:/dev/sda8 DeviceMajor:8 DeviceMinor:8 Capacity:12042240 Type:vfs Inodes:4096 HasInodes:true} {Device:/dev/sda1 DeviceMajor:8 DeviceMinor:1 Capacity:101241290752 Type:vfs Inodes:6258720 HasInodes:true} {Device:/dev/shm DeviceMajor:0 DeviceMinor:19 Capacity:3911983104 Type:vfs Inodes:955074 HasInodes:true} {Device:/run DeviceMajor:0 DeviceMinor:21 Capacity:3911983104 Type:vfs Inodes:955074 HasInodes:true} {Device:/sys/fs/cgroup DeviceMajor:0 DeviceMinor:22 Capacity:3911983104 Type:vfs Inodes:955074 HasInodes:true} {Device:/etc/machine-id DeviceMajor:0 DeviceMinor:21 Capacity:3911983104 Type:vfs Inodes:955074 HasInodes:true} {Device:/mnt/disks DeviceMajor:0 DeviceMinor:38 Capacity:262144 Type:vfs Inodes:955074 HasInodes:true} {Device:overlayfs_0-43 DeviceMajor:0 DeviceMinor:43 Capacity:3911983104 Type:vfs Inodes:955074 HasInodes:true} {Device:/var/lib/cloud DeviceMajor:0 DeviceMinor:45 Capacity:1048576 Type:vfs Inodes:955074 HasInodes:true} {Device:/dev/root DeviceMajor:253 DeviceMinor:0 Capacity:2054897664 Type:vfs Inodes:127488 HasInodes:true}] DiskMap:map[253:0:{Name:dm-0 Major:253 Minor:0 Size:2087714816 Scheduler:none} 8:0:{Name:sda Major:8 Minor:0 Size:107374182400 Scheduler:mq-deadline} 9:0:{Name:md0 Major:9 Minor:0 Size:0 Scheduler:none}] NetworkDevices:[{Name:eth0 MacAddress:42:01:0a:28:00:03 Speed:-1 Mtu:1460}] Topology:[{Id:0 Memory:7823970304 HugePages:[{PageSize:2048 NumPages:0}] Cores:[{Id:0 Threads:[0 1] Caches:[{Size:32768 Type:Data Level:1} {Size:32768 Type:Instruction Level:1} {Size:262144 Type:Unified Level:2}] SocketID:0}] Caches:[{Size:57671680 Type:Unified Level:3}]}] CloudProvider:GCE InstanceType:n1-standard-2 InstanceID:5238975090913384389}
I1113 15:43:26.950497    1338 manager_no_libpfm.go:28] **cAdvisor is build without cgo** and/or libpfm support. Perf event counters are not available.
I1113 15:43:26.950882    1338 manager.go:229] Version: {KernelVersion:5.4.49+ ContainerOsVersion:Container-Optimized OS from Google DockerVersion:Unknown DockerAPIVersion:Unknown CadvisorVersion: CadvisorRevision:}

I guess kubelet can't load in cAdvisor the NVML library.

**Comment by ruiwen-zhao** [comment]
@AlexeyPerevalov I have looked into this issue actually. This is caused by a recent change on cadvisor: https://github.com/google/cadvisor/pull/2419/files#diff-bd4495842459f1c8cdcb90c14dc169abdcd738596d7e7710daa2f9a9aa7a7ba5

We used to return a real nvidiaManager but now only a Noop one if the NVML lib is not there. And currently the NVML lib is loaded by daemonset, which will not there when the kubelet starts. 

This PR should fix it: https://github.com/google/cadvisor/pull/2732
